### ISSUE TITLE
Spike to evaluate feasibility and scope of factoring rewiring logic out of `ModuleTranslator`

### ISSUE THREAD

**Issue by shonfeder** [informalsystems/apalache#2479]
The trickiest part or our [quint conversion milestone](https://github.com/informalsystems/apalache/milestone/74) is converting the operators 
that require rewiring with specially defined implementations (e.g. [`SelectSeq`](https://github.com/informalsystems/apalache/blob/91ad7b4edff727f1e8579ca5b2fac1ac2c41bb15/src/tla/__rewire_sequences_in_apalache.tla#L92-L105). See #2437.

The only reason we even need to worry about this in the Quint 
conversion is that the rewiring operation that should ideally
be a simple mapping `TlaModule -> TlaModule` is instead deeply
intertwined with the (pretty hard to follow, imo) translation
from Sany's data types into the Apalache IR. This architectural
issue impacts current functionality, irrespective of Quint, since
input supplied as JSON cannot be rewired.

The task set out in this issue, which @thpani and I discussed 
on Friday, is to evaluate how feasible it would be to factor out
the rewiring, so it can be used for input arriving from any source.

Here are some notes on the general problem and a possible solution we've
discussed:

Suppose (contrary to fact) that we had the following representation of a
parsed TLA module:

``` scala
// A TLA spec module parsed onto Apalache's IR
class TlaModule {...}

// A TlaModule which has been enriched by having its declaraions rewired
// with the contents of the /apalache-core/src/tla/__rewire*.tla moduls
case object RewiredTlaModule extends TlaModule
```

The core problem: we'd like all inputs to construct a
`RewiredTlaModule`, but the current implementation only produces the
weaker `TlaModule` in general.

Abstracting and simplifying to the happy path, our [current
plumbing](https://github.com/informalsystems/apalache/blob/main/passes/src/main/scala/at/forsyte/apalache/tla/passes/imp/SanyParserPassImpl.scala#L121)
is like this:

``` scala
// The type of inputs indexed by format
type Format = Tla | Qnt | Json
class SourceOption(format: Format)

class SanyParserPassImpl {

  val loadFromTlaSource  : SourceOption => RewiredTlaModule
  val loadFromJsonSource : SourceOption => TlaModule

  val parseSource : SourceOption => TlaModule =
    src => s.format match {
      case Tla          => loadFromTlaSource(src)
      case (Qnt | Json) => loadFromJsonSource(src)
    }
}
```

`loadFromTlaSource` can take us from `SourceOption` to
`RewiredTlaModule`, but since `loadFromJsonSource` goes
`SourceOption => TlaModule`, we can only rely on `parseModule` to give
us a `TlaModule`.

The suspected solution to our problem if to factor out the enrichment
internal to `loadFromTlaSource`, so that we can we have a shape like:

``` scala
// The type of inputs indexed by format
type Format = Tla | Qnt | Json
class SourceOption(format: Format)

class SanyParserPassImpl {

  val loadFromTlaSource  : SourceOption => TlaModule
  val loadFromJsonSource : SourceOption => TlaModule

  val rewireModule : TlaModule => RewiredTlaModule

  val parseSource : SourceOption => RewiredTlaModule =
    src => s.format match {
      case Tla          => rewireModule(loadFromTlaSource(src))
      case (Qnt | Json) => rewireModule(loadFromJsonSource(src))
    }
}
```

Our aim it to construct the mapping
`rewireModule : TlaModule => RewiredTlaModule`.

The complication arising from our current architecture follows from the
fact that `loadFromTlaSource` has an unrepresented input: namely the
`/apalache-core/src/tla/*.tla` files which are read off of the file
system. These are pulled in via side-effects in
[`SanyImproter.loadFromFile`](https://github.com/informalsystems/apalache/blob/91ad7b4edff727f1e8579ca5b2fac1ac2c41bb15/tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala#L66-L100).

Representing this in our spec, we get

``` scala
// The type of inputs indexed by format
type Format = Tla | Qnt | Json
class SourceOption(format: Format)

class SanyParserPassImpl {

  val loadFromTlaSource  : SourceOption => TlaModule
  val loadFromJsonSource : SourceOption => TlaModule

  val tlaAssets : Seq[TlaModule] = // Read from disk

  val rewireModule : (Seq[TlaModuls], TlaModule) => RewiredTlaModule

  val parseSource : SourceOption => RewiredTlaModule =
    src => s.format match {
      case Tla          => rewireModule(tlaAssests, loadFromTlaSource(src))
      case (Qnt | Json) => rewireModule(tlaAssests, loadFromJsonSource(src))
    }
}
```

One obstacle to the solution proposed by `rewireModule` is that the only
way we have to get from a `.tla` file to a `TlaModule` is via Sany.

In theory, this shouldn't be much of an obstacle: given a function
`loadStandardModules : () => Seq[TlaModule]` that abstracts the logic
from
[`SanyImproter.loadFromFile`](https://github.com/informalsystems/apalache/blob/91ad7b4edff727f1e8579ca5b2fac1ac2c41bb15/tla-io/src/main/scala/at/forsyte/apalache/tla/imp/SanyImporter.scala#L66-L100)
to load just the assets in `/apalache-core/src/tla/*.tla`, we should be
able to apply `rewireModule`. In fact, the current situation poses two
more significant challenges:

1.  Sany must have access to all our module dependencies at parse time,
    and we currently don't make any distinction between our extended
    standard library modules (like `Apalache`) and the rewiring modules.
    So until we can factor out loading the rewiring modules from loading
    the standard modules, we can't use `rewireModule` in the `TLA` case.
2.  The much bigger problem is that the actual rewiring is totally
    enmeshed with our translation from Sany's representation into the
    Apalache IR: we have nothing resembling `rewireModule` in the guts
    of the current `loadFromTlaSource`. So even once we've figured out a
    way to get our `tlaAssets`, we will still have to do significant
    work to recover the rewiring operation from the tangled guts of the
    [ModuleTranslator](https://github.com/informalsystems/apalache/blob/main/tla-io/src/main/scala/at/forsyte/apalache/tla/imp/ModuleTranslator.scala#L18).

We get a sense for how intertwined the rewiring currently is by tracing
out how we go from the identification of units needing rewiring in
<https://github.com/informalsystems/apalache/blob/91ad7b4edff727f1e8579ca5b2fac1ac2c41bb15/tla-io/src/main/scala/at/forsyte/apalache/tla/imp/ModuleTranslator.scala#L118-L126>
to the actual location where the replacements occur in
<https://github.com/informalsystems/apalache/blob/91ad7b4edff727f1e8579ca5b2fac1ac2c41bb15/tla-io/src/main/scala/at/forsyte/apalache/tla/imp/ExprOrOpArgNodeTranslator.scala#L155-L165>

afaict, there is nothing that necessitates the lack of clean separation
between translation out of Sany and the rewiring, but I could be missing
some essential complexity.

**Comment by thpani** [comment]
Quick update: I believe we're better off postponing this until a more substantial use-case comes up.

Benefits / drawbacks as I currently see them:

Benefits:
- Would save duplicating some logic between TLA+ rewiring modules and the Quint translator
- Could specify more complex translations in TLA+, rather than using the typed builder

Drawbacks:
- Not clear that we'd want to rely on TLA+ community modules for our Quint translation (see #2437), they might diverge in the future.
- The JSON import use case is limited: JSON input is currently self-contained and cannot refer to other modules.
- Untangling of current rewiring logic is likely tedious.
- Unclear how to translate the current rewiring modules. They don't parse out-of-the box.
- Parsing the rewiring modules through Apalache would create a circular build dependency.

After sync discussion with the team, we decided that I'll write a translation for one of the operators in #2437) in code, then follow-up on how to proceed.
